server/src/main/java/io/druid/server/coordinator/rules/BroadcastDistributionRule.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.server.coordinator.rules;
+
+import com.metamx.emitter.EmittingLogger;
+import io.druid.server.coordinator.CoordinatorStats;
+import io.druid.server.coordinator.DruidCoordinator;
+import io.druid.server.coordinator.DruidCoordinatorRuntimeParams;
+import io.druid.server.coordinator.ServerHolder;
+import io.druid.timeline.DataSegment;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public abstract class BroadcastDistributionRule implements Rule
+{
+  private static final EmittingLogger log = new EmittingLogger(BroadcastDistributionRule.class);
+
+  @Override
+  public CoordinatorStats run(
+      DruidCoordinator coordinator, DruidCoordinatorRuntimeParams params, DataSegment segment
+  )
+  {
+    // Find servers which holds the segments of co-located data source
+    final Set<ServerHolder> loadServerHolders = new HashSet<>();
+    final Set<ServerHolder> dropServerHolders = new HashSet<>();
+    final List<String> colocatedDataSources = getColocatedDataSources();
+    if (colocatedDataSources == null || colocatedDataSources.isEmpty()) {
+      loadServerHolders.addAll(params.getDruidCluster().getAllServers());
+    } else {
+      params.getDruidCluster().getAllServers().forEach(
+          eachHolder -> {
+            if (colocatedDataSources.stream()
+                                    .anyMatch(source -> eachHolder.getServer().getDataSource(source) != null)) {
+              loadServerHolders.add(eachHolder);
+            } else if (eachHolder.isServingSegment(segment)) {
+              dropServerHolders.add(eachHolder);
+            }
+          }
+      );
+    }
+
+    final CoordinatorStats stats = new CoordinatorStats();
+
+    return stats.accumulate(assign(loadServerHolders, segment))
```

Comment:
I think broadcasted segments should be regarded as same as other segments. Every types of queries can be queried on broadcasted segments, but the broadcast information is required by only joins.

The information of which sources are broadcasted or not is required by brokers and historicals (and realtimes in the future). Brokers decides a given query should be distributed to which nodes, and this is determined by selecting the nodes holding the segments of non-broadcasted tables. Historicals, as proposed in #4032, join broadcasted segments first and then subsequently join the result with non-broadcasted segments in parallel.

I'm currently thinking that brokers are able to figure out which tables are broadcasted by looking at BroadcastDistributionRules, and they will add this information to QueryContext for historicals. If this works, broadcast segments can be regarded as normal. 

What do you think?